apps/examples/testcase/ta_tc/audio/utc/utc_audio_main.c

```diff
@@ -638,43 +635,46 @@ static void utc_audio_pcm_readi_p(void)
 {
 	int fd;
 	char input_str[16];
-	ssize_t size;
-	ssize_t bytes;
-	int frames_read = 0;
-	unsigned int total_frames_read = 0;
-	unsigned int rate;
-	int remain;
 	int ret;
 	char *str;
 
-	size = pcm_get_buffer_size(g_pcm);
-	rate = pcm_get_rate(g_pcm);
-	bytes = rate * (pcm_format_to_bits(pcm_get_format(g_pcm)) / 8) * pcm_get_channels(g_pcm) * AUDIO_RECORD_DURATION;
+	char *buffer;
+	unsigned int size;
+	unsigned int frames_read;
+	unsigned int total_frames_read;
+	unsigned int bytes_per_frame;
+	int capturing = 1;
+
+	size = pcm_frames_to_bytes(g_pcm, pcm_get_buffer_size(g_pcm));
+	buffer = malloc(size);
+	TC_ASSERT_NEQ_CLEANUP("pcm_readi", buffer, NULL, clean_all_data(0, 0));
 
-	g_record_buffer = (char *)malloc(bytes);	//rate * bit * channels * duration
-	TC_ASSERT_NEQ("pcm_readi_p", g_record_buffer, NULL);
+	fd = open(AUDIO_TEST_FILE, O_RDWR | O_CREAT);
+	TC_ASSERT_GT_CLEANUP("pcm_readi", fd, 0, clean_all_data(0, buffer));
 
-	remain = pcm_bytes_to_frames(g_pcm, bytes);
-	printf("Record will be start for 3s, press any key to start(Total frame: %d, %d bytes)\n", remain, bytes);
+	printf("Record will be start for 3s, press any key to start\n");
 	fflush(stdout);
 	str = gets(input_str);
-	TC_ASSERT_NEQ_CLEANUP("pcm_readi", str, NULL, free(g_record_buffer));
+	TC_ASSERT_NEQ_CLEANUP("pcm_readi", str, NULL, clean_all_data(fd, buffer));
 
-	while (remain > 0) {
-		frames_read = pcm_readi(g_pcm, (void *)&g_record_buffer[pcm_frames_to_bytes(g_pcm, total_frames_read)], remain);
+	bytes_per_frame = pcm_frames_to_bytes(g_pcm, 1);
+	total_frames_read = 0;
+	frames_read = 0;
+	while (capturing) {
+		frames_read = pcm_readi(g_pcm, buffer, pcm_get_buffer_size(g_pcm));
 		total_frames_read += frames_read;
-		remain -= frames_read;
-	}
-	printf("Record done. Writing file...\n");
+		if ((total_frames_read / AUDIO_DEFAULT_RATE) >= AUDIO_RECORD_DURATION) {
+			capturing = 0;
+		}
 
-	pcm_close(g_pcm);
-	fd = open(AUDIO_TEST_FILE, O_RDWR | O_CREAT);
-	ret = write(fd, g_record_buffer, bytes);
-	TC_ASSERT_GT_CLEANUP("pcm_reaid_p", ret, 0, close(fd));
-	close(fd);
+		ret = write(fd, buffer, bytes_per_frame * frames_read);
+		TC_ASSERT_EQ_CLEANUP("pcm_readi", ret, (bytes_per_frame * frames_read), clean_all_data(fd, buffer));
```

Comment:
if remained data is smaller than buffer size, ret is not same as tyte_per_frame * frames_read.
0 is correct.